zipline/pipeline/data/dataset.py

```diff
@@ -497,3 +500,229 @@ class CompanyMetadata(DataSet):
+
+    def __new__(cls, name, bases, dict_):
+        columns = {}
+        for k, v in dict_.copy().items():
+            if isinstance(v, Column):
+                columns[k] = v
+                del dict_[k]
+
+        is_base_class = bases == (cls._base_marker,)
+        if is_base_class:
+            bases = (object,)
+        self = super(MultiDimensionalDataSetMeta, cls).__new__(
+            cls,
+            name,
+            bases,
+            dict_,
+        )
+
+        if not is_base_class:
+            self.free_axes = free_axes = OrderedDict([
+                (k, frozenset(v))
+                for k, v in OrderedDict(self.free_axes).items()
+            ])
+            if not free_axes:
+                raise ValueError(
+                    'MultiDimensionalDataSet must be defined with non-empty'
+                    ' free_axes',
+                )
+
+            class Slice(self._SliceType):
+                parent_multidimensional_dataset = self
+
+                ndim = self.ndim
+                domain = self.domain
+
+                locals().update(columns)
+
+            Slice.__name__ = '%sBaseSlice' % self.__name__
+            self._SliceType = Slice
+
+        # each type gets a unique cache
+        self._slice_cache = {}
+        return self
+
+
+_base = with_metaclass(
+    MultiDimensionalDataSetMeta,
+    MultiDimensionalDataSetMeta._base_marker,
+)
+
+
+class MultiDimensionalDataSetSlice(DataSet):
+    """Marker type for slices of a
+    :class:`zipline.pipeline.data.dataset.MultiDimensionalDataSet` objects
+    """
+
+
+class MultiDimensionalDataSet(_base):
+    """
+    Base class for Pipeline multi-dimensional datasets.
+
+    A multi-dimensional dataset represents data where the unique identifier for
+    a particular value requires more than asset and date labels. A
+    multi-dimensional dataset may be thought of as a collection of
+    :class:`~zipline.pipeline.data.DataSet` objects with the same columns,
+    domain, and ndim.
+
+    ``MultiDimensionalDataSet`` objects have an extra field called the
+    ``free_axes``. The ``free_axes`` field describes the labels that are not
+    asset or date. The ``free_axes`` are represented as an ordered dictionary
+    where the keys are the axis name, and the values are a set of unique values
+    along that axis.
+
+    To use a ``MultiDimensionalDataSet``, one must "fix" all of the free axes.
+    The :meth:`~zipline.pipeline.data.dataset.MultiDimensionalDataSet.slice`
+    method is used to create a dataset where all rows have the same values in
+    the free axes. For example, given a ``MultiDimensionalDataSet``:
+
+    .. code-block:: python
+
+       class SomeDataSet(MultiDimensionalDataSet):
+           free_axes = [
+               ('axis_0', {'a', 'b', 'c'}),
+               ('axis_1', {'d', 'e', 'f'}),
+           ]
+
+           column_0 = Column('f8')
+           column_1 = Column('?')
+
+    This represents a table with the following columns:
```

Comment:
Except for our weird handling of fundamentals, everything does map pretty closely to a table where each column shares the same coordinates. This documentations is also for dataset implementers, not consumers.

Reply:
At least right now the docs for `DataSet` are also on Quantopian, because they're part of the public API. We're likely going to want user-facing documentation for this type eventually as well, at least for `slice` and `aggregate`, but also generally for the concept of multi-dimensional datasets.

Reply:
What do you think about changing this from "This represents a table" to "This dataset might represent a table"?